<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Workflow for NetPathMiner • NetPathMiner</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Workflow for NetPathMiner">
<meta property="og:description" content="NetPathMiner">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">NetPathMiner</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.35.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/NPMVignette.html">Workflow for NetPathMiner</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/ahmohamed/NetPathMiner/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Workflow for NetPathMiner</h1>
                        <h4 data-toc-skip class="author">Ahmed
Mohamed</h4>
            <address class="author_afil">
      QIMR Berghofer Medical Research Institute, 300 Herston Road,
Herston QLD 4006, Australia<br><h4 data-toc-skip class="date">2023-02-20</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/ahmohamed/NetPathMiner/blob/HEAD/vignettes/NPMVignette.Rmd" class="external-link"><code>vignettes/NPMVignette.Rmd</code></a></small>
      <div class="hidden name"><code>NPMVignette.Rmd</code></div>

    </address>
</div>

    
    
<div class="section level2">
<h2 id="version-info">Version Info<a class="anchor" aria-label="anchor" href="#version-info"></a>
</h2>
<p>
<strong>R version</strong>: R Under development (unstable) (2023-02-16
r83857) <br><strong>Bioconductor version</strong>: 3.17 <br><strong>Package version</strong>: 1.35.5
</p>
</div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>NetPathMiner implements a flexible module-based process flow for
network path mining and visualization, which can be fully integrated
with user-customized functions. It supports construction of various
types of genome scale networks from three different pathway file
formats, enabling its utility to most common pathway databases. In
addition, NetPathMiner provides different visualization techniques to
facilitate the analysis of even thousands of output paths.</p>
<p>This document provides a general overview of the functionalities
presented in NetPathMiner (NPM) package. Below, we provide a
step-by-step tutorial starting by installation instructions followed by
a guide on how to use the package functions perform different network
analyses.</p>
<p>To report bugs and arising issues, please visit <a href="https://github.com/ahmohamed/NetPathMiner" class="external-link uri">https://github.com/ahmohamed/NetPathMiner</a></p>
</div>
<div class="section level2">
<h2 id="installation-instructions">Installation Instructions<a class="anchor" aria-label="anchor" href="#installation-instructions"></a>
</h2>
<div class="section level3">
<h3 id="system-prerequisites">System Prerequisites<a class="anchor" aria-label="anchor" href="#system-prerequisites"></a>
</h3>
<p>NPM depends on libxml2 and libSBML to process pathway
files.Installation or running certain functions MIGHT fail if these
prerequisite libraries are not available. Please read through the
following instructions.</p>
<div class="section level5">
<h5 id="prerequisites-for-unix-users-linux-and-mac-os">Prerequisites for Unix users (Linux and Mac OS)<a class="anchor" aria-label="anchor" href="#prerequisites-for-unix-users-linux-and-mac-os"></a>
</h5>
<div class="section level6">
<h6 id="installing-libxml2">Installing libxml2<a class="anchor" aria-label="anchor" href="#installing-libxml2"></a>
</h6>
<p>Make sure your system has library libxml2 installed. In Mac OSX 10.6
or later, libxml2 are built in. For Linux users also, this is almost
always the case, however, developing headers may be missing. To install
libxml2 and the headers:</p>
<pre class="shell"><code>    sudo apt-get install libxml2
    sudo apt-get install libxml2-dev</code></pre>
</div>
<div class="section level6">
<h6 id="installing-libsbml">Installing libSBML<a class="anchor" aria-label="anchor" href="#installing-libsbml"></a>
</h6>
<p>Installing libSBML for Unix users is optional. However, NetPathMiner
will not be able to process SBML files. If you will not use SBML
functions, you can skip this part.</p>
<p>From the website of libSBML <a href="http://sbml.org/Software/libSBML" class="external-link uri">http://sbml.org/Software/libSBML</a>, you can directly
download the binaries suitable for your system from
<code>Download libSBML</code> link. You can follow the installation
instructions on the website.</p>
</div>
</div>
<div class="section level5">
<h5 id="prerequisites-for-windows-users">Prerequisites for Windows users<a class="anchor" aria-label="anchor" href="#prerequisites-for-windows-users"></a>
</h5>
<p>If you are installing the package through Bioconductor, you don’t
have to install external libraries. However, currently the Bioconductor
version for Windows doesn’t support SBML processing. Alternatively, we
have prepared all dependencies in a tar file, downloadable from <a href="https://github.com/ahmohamed/NPM_dependencies" class="external-link uri">https://github.com/ahmohamed/NPM_dependencies</a> . Please
download the file and place in in the home directory of R (type <code>R
RHOME</code> in command prompt to locate it), before installation.</p>
<p>Unless you want to use customized libraries, you can skip the rest of
this section.To use customized libraries, you have to compile them and
provide them to R at the time of installation. This is not a trivial
task, please be sure you really need these custom libraries.</p>
<div class="section level6">
<h6 id="installing-libxml2-1">Installing libxml2<a class="anchor" aria-label="anchor" href="#installing-libxml2-1"></a>
</h6>
<p>NetPathMiner expects an enviroment variable <code>LIB_XML</code> or
<code>LIB_XML2</code> pointing to directory where libxml2 is installed.
This directory should have both the compiled library (DLL file) and the
header files.</p>
<p>You can download libxml2.dll from <a href="http://sourceforge.net/projects/gnuwin32/files/libxml/" class="external-link uri">http://sourceforge.net/projects/gnuwin32/files/libxml/</a>
among other sources. Please, place it in a <code>bin</code> folder under
the installation directory.</p>
<p>You will need also the header files, which can be obtained from
NPM_dependecies.tar file. After extracing it, copy the include directory
to the installation directory.</p>
<p>Finally, set the <code>LIB_XML2</code> variable to point to the
installation directory, which should now contain dll files inder
<code>bin</code> and header files under <code>include</code>.</p>
</div>
<div class="section level6">
<h6 id="installing-libsbml-1">Installing libSBML<a class="anchor" aria-label="anchor" href="#installing-libsbml-1"></a>
</h6>
<p>Since libSBML is a C++ libraries, it needs to be compiled using GCC
compiler. Unforturantely, there is no binary version for Windows
comipled with GCC. To use libSBML, you need to build it from source.</p>
<p>First, dowload source package from <a href="http://sourceforge.net/projects/sbml/files/libsbml/" class="external-link uri">http://sourceforge.net/projects/sbml/files/libsbml/</a> ,
extract it. You will need also MinGW <a href="http://www.mingw.org/" class="external-link uri">http://www.mingw.org/</a> or the 64-bit version <a href="http://mingw-w64.sourceforge.net/" class="external-link uri">http://mingw-w64.sourceforge.net/</a> depending on your
system. Add <code>mingw/bin</code> to your PATH, by editing eviroment
variables.</p>
<p>Second, you need CMake <a href="http://www.cmake.org/" class="external-link uri">http://www.cmake.org/</a> . You can follow the instructions
at <a href="http://sbml.org/Software/libSBML/docs/java-api/libsbml-installation.html#windows-configuring" class="external-link uri">http://sbml.org/Software/libSBML/docs/java-api/libsbml-installation.html#windows-configuring</a>
, however, choose “MinGW Makefiles” instead of “Visual Studio 10”.</p>
<p>After finishing the CMake step, use the MinGW’s <code>make.exe</code>
to compile libSBML. Copy the dependencies you used during the
compilation to the <code>bin</code> directory. Set the enviroment
variable <code>LIB_SBML</code> to point the installation directory,
which should now contain dll files inder <code>bin</code> and header
files under <code>include</code></p>
</div>
</div>
<div class="section level4">
<h4 id="r-package-dependencies">R Package dependencies<a class="anchor" aria-label="anchor" href="#r-package-dependencies"></a>
</h4>
<p>NetPathMiner depends on package igraph to represent network objects.
Installing igraph is required for the package to work. You will also
need devtools package to install directly from github. NetPathMiner
suggests package rBiopaxParser to process BioPAX files and RCurl to
download annotations from the web. NetPathMiner can still work without
installing the suggested packages, but you will not be able to use the
aforementioned functionalities.</p>
<div class="section level6">
<h6 id="igraph">igraph<a class="anchor" aria-label="anchor" href="#igraph"></a>
</h6>
<p>Package igraph is available at CRAN. To install it call:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"igraph"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level6">
<h6 id="devtools">devtools<a class="anchor" aria-label="anchor" href="#devtools"></a>
</h6>
<p>Package devtools is available at CRAN. For Windows, this seems to
depend on having Rtools for Windows installed. You can download and
install this from: <a href="http://cran.r-project.org/bin/windows/Rtools/" class="external-link uri">http://cran.r-project.org/bin/windows/Rtools/</a></p>
<p>To install R package devtools call:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level6">
<h6 id="rcurl">RCurl<a class="anchor" aria-label="anchor" href="#rcurl"></a>
</h6>
<p>For Unix users, make sure your Linux has library libcurl installed.
Check out:</p>
<pre class="shell"><code>    locate libcurl
    locate curl-config</code></pre>
<p>If these are not found (usually the developer version is missing),
most Linux users will be able to fix this by running:</p>
<pre class="shell"><code>    sudo apt-get install libcurl4-openssl-dev</code></pre>
<p>You will now be able to install R package RCurl. In R console:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"RCurl"</span><span class="op">)</span></span></code></pre></div>
<p>If you encounter other problems check out <a href="http://www.omegahat.org/RCurl/FAQ.html" class="external-link uri">http://www.omegahat.org/RCurl/FAQ.html</a></p>
</div>
<div class="section level6">
<h6 id="rbiopaxparser">rBiopaxParser<a class="anchor" aria-label="anchor" href="#rbiopaxparser"></a>
</h6>
<p>Package rBiopaxParser is available on Bioconductor. For installation
instructions check out <a href="http://www.bioconductor.org/packages/release/bioc/html/rBiopaxParser.html" class="external-link uri">http://www.bioconductor.org/packages/release/bioc/html/rBiopaxParser.html</a>
or call:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>    <span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span>    <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"rBiopaxParser"</span><span class="op">)</span></span></code></pre></div>
<p>to install it right away.</p>
</div>
</div>
<div class="section level4">
<h4 id="netpathminer-installation">NetPathMiner Installation<a class="anchor" aria-label="anchor" href="#netpathminer-installation"></a>
</h4>
<p>If everything went well you will be able to install the NetPathMiner
package.</p>
<div class="section level5">
<h5 id="from-bioconductor">From Bioconductor:<a class="anchor" aria-label="anchor" href="#from-bioconductor"></a>
</h5>
<p>In R console, type:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>    <span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span>    <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"NetPathMiner"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="from-github-using-devtools">From GitHub using devtools:<a class="anchor" aria-label="anchor" href="#from-github-using-devtools"></a>
</h5>
<p>In R console, type:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/" class="external-link">devtools</a></span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span>repo<span class="op">=</span><span class="st">"NetPathMiner"</span>, username<span class="op">=</span><span class="st">"ahmohamed"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="getting-started">Getting Started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<p>First, let’s load the library and the example data set.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ahmohamed/NetPathMiner" class="external-link">NetPathMiner</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="database-extraction">Database Extraction<a class="anchor" aria-label="anchor" href="#database-extraction"></a>
</h2>
<p>Here we create a network from a pathway file. Pathway files can be
downloaded from a verity of databases, like (KEGG)[<a href="http://www.kegg.jp/kegg/pathway.html" class="external-link uri">http://www.kegg.jp/kegg/pathway.html</a>], (Reactome)[<a href="http://www.reactome.org/" class="external-link uri">http://www.reactome.org/</a>], (Pathway Interaction Database
(PID))[<a href="http://pid.nci.nih.gov/" class="external-link uri">http://pid.nci.nih.gov/</a>] and (BioModels)[<a href="http://www.ebi.ac.uk/biomodels-main/" class="external-link uri">http://www.ebi.ac.uk/biomodels-main/</a>].</p>
<p>NPM supports processing KGML, SBML and BioPAX. Different databases
export pathway information in different formats. Depending on the format
you are using, you can choose the corresponding NPM function.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">graph</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/KGML2igraph.html">KGML2igraph</a></span><span class="op">(</span>filename <span class="op">=</span> <span class="va">file</span><span class="op">)</span></span>
<span><span class="va">graph</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SBML2igraph.html">SBML2igraph</a></span><span class="op">(</span>filename <span class="op">=</span> <span class="va">file</span><span class="op">)</span></span></code></pre></div>
<p>Note that SBML2igraph will not work unless you had libSBML during
installation. For BioPAX format, <code>rBiopaxParser</code> package is
needed.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/frankkramer-lab/rBiopaxParser" class="external-link">rBiopaxParser</a></span><span class="op">)</span></span>
<span><span class="va">biopax</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/rBiopaxParser/man/readBiopax.html" class="external-link">readBiopax</a></span><span class="op">(</span><span class="va">file</span><span class="op">)</span></span>
<span><span class="va">graph</span> <span class="op">&lt;-</span> <span class="fu">BioPAX2igraph</span><span class="op">(</span>biopax <span class="op">=</span> <span class="va">biopax</span><span class="op">)</span></span></code></pre></div>
<p>If you want to create a genome-scale network, you may want to process
multiple files into a single network. To do that, you can either provide
a list of files.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">graph</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/KGML2igraph.html">KGML2igraph</a></span><span class="op">(</span>filename <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">file1</span>, <span class="va">file2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>or input the directory containing the files.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">graph</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/KGML2igraph.html">KGML2igraph</a></span><span class="op">(</span>filename <span class="op">=</span> <span class="st">"."</span><span class="op">)</span></span></code></pre></div>
<p>If you are processing SBML or BioPAX files, you can specify which
annotation attributes to extract.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Extract all MIRIAM identifiers from an SBML file.</span></span>
<span><span class="va">graph</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SBML2igraph.html">SBML2igraph</a></span><span class="op">(</span>filename <span class="op">=</span> <span class="va">file</span>, miriam <span class="op">=</span> <span class="st">"all"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract only miram.go identifiers from a BioPAX file.</span></span>
<span><span class="va">graph</span> <span class="op">&lt;-</span> <span class="fu">BioPAX2igraph</span><span class="op">(</span>biopax <span class="op">=</span> <span class="va">biopax</span>, miriam <span class="op">=</span> <span class="st">"go"</span><span class="op">)</span></span></code></pre></div>
<p>The above command gives us a bipartite metabolic network. You may be
interested in protein-protein interaction, and you would like to get a
network in which genes are vertices, and edges represent relationships.
In NPM, you can do that by:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">graph</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/KGML2igraph.html">KGML2igraph</a></span><span class="op">(</span>filename <span class="op">=</span> <span class="va">file</span>, parse.as <span class="op">=</span> <span class="st">"signaling"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">graph</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/KGML2igraph.html">KGML2igraph</a></span><span class="op">(</span>filename <span class="op">=</span> <span class="va">file</span>, parse.as <span class="op">=</span> <span class="st">"signaling"</span>,</span>
<span>    expand.complexes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>For this tutorial, we will use an An example metabolic network of
Carbohydrate metabolism extracted from SBML file from Reactome
database.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"ex_sbml"</span><span class="op">)</span></span>
<span><span class="va">graph</span> <span class="op">&lt;-</span> <span class="va">ex_sbml</span></span>
<span><span class="va">graph</span></span></code></pre></div>
<pre><code><span><span class="co">## IGRAPH 1981d4c DN-- 432 662 -- </span></span>
<span><span class="co">## + attr: source (g/c), type (g/c), name (v/c), attr (v/x), reactions</span></span>
<span><span class="co">## | (v/l), shape (v/c), color (v/c), stoichiometry (e/n)</span></span>
<span><span class="co">## + edges from 1981d4c (vertex names):</span></span>
<span><span class="co">##  [1] reaction_71850  -&gt;species_113592  reaction_71850  -&gt;species_29728  </span></span>
<span><span class="co">##  [3] reaction_71496  -&gt;species_75970   reaction_71496  -&gt;species_29578  </span></span>
<span><span class="co">##  [5] reaction_2046222-&gt;species_2046166 reaction_2046222-&gt;species_742343 </span></span>
<span><span class="co">##  [7] reaction_2162229-&gt;species_2162232 reaction_2162229-&gt;species_2162231</span></span>
<span><span class="co">##  [9] reaction_2162226-&gt;species_1678868 reaction_2162226-&gt;species_1678743</span></span>
<span><span class="co">## [11] reaction_70403  -&gt;species_70113   reaction_2162227-&gt;species_1678868</span></span>
<span><span class="co">## [13] reaction_2162227-&gt;species_2162223 reaction_2162225-&gt;species_1678743</span></span>
<span><span class="co">## + ... omitted several edges</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="handling-annotation-attributes">Handling Annotation Attributes<a class="anchor" aria-label="anchor" href="#handling-annotation-attributes"></a>
</h2>
<p>Once we have our network, we can use igraph functions to explore it.
First, we view vertices and edges using V() and E() functions
respectively.</p>
<p>Network vertices:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/pkg/igraph/man/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span> <span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## + 6/432 vertices, named, from 1981d4c:</span></span>
<span><span class="co">## [1] reaction_71850   reaction_71496   reaction_2046222 reaction_2162229</span></span>
<span><span class="co">## [5] reaction_2162226 reaction_70403</span></span></code></pre>
<p>Edge Vertices:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/pkg/igraph/man/E.html" class="external-link">E</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span> <span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## + 6/662 edges from 1981d4c (vertex names):</span></span>
<span><span class="co">## [1] reaction_71850  -&gt;species_113592  reaction_71850  -&gt;species_29728  </span></span>
<span><span class="co">## [3] reaction_71496  -&gt;species_75970   reaction_71496  -&gt;species_29578  </span></span>
<span><span class="co">## [5] reaction_2046222-&gt;species_2046166 reaction_2046222-&gt;species_742343</span></span></code></pre>
<p>Reaction vertices only:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/pkg/igraph/man/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span><span class="op">[</span> <span class="va">reactions</span> <span class="op">]</span> <span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## + 6/432 vertices, named, from 1981d4c:</span></span>
<span><span class="co">## [1] reaction_71850   reaction_71496   reaction_2046222 reaction_2162229</span></span>
<span><span class="co">## [5] reaction_2162226 reaction_70403</span></span></code></pre>
<p>All vertex annotation attributes are stored in “attr” attribute. To
view the annotation for a certain vertex, you can index it by name.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span><span class="op">[</span> <span class="st">"reaction_71850"</span> <span class="op">]</span><span class="op">$</span><span class="va">attr</span></span></code></pre></div>
<pre><code><span><span class="co">## [[1]]</span></span>
<span><span class="co">## [[1]]$name</span></span>
<span><span class="co">## [1] "1,3-bisphospho-D-glycerate + ADP &lt;=&gt; 3-phospho-D-glycerate + ADP"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[1]]$reversible</span></span>
<span><span class="co">## [1] FALSE</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[1]]$reactants</span></span>
<span><span class="co">## [1] "species_29800" "species_29370"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[1]]$reactant.stoichiometry</span></span>
<span><span class="co">## [1] 1 1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[1]]$products</span></span>
<span><span class="co">## [1] "species_29728"  "species_113592"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[1]]$product.stoichiometry</span></span>
<span><span class="co">## [1] 1 1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[1]]$kinetics</span></span>
<span><span class="co">## named list()</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[1]]$genes</span></span>
<span><span class="co">## [1] "phosphoglycerate kinase 1 complex [cytosol]"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[1]]$compartment</span></span>
<span><span class="co">## [1] "compartment_70101"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[1]]$compartment.name</span></span>
<span><span class="co">## [1] "cytosol"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[1]]$pathway</span></span>
<span><span class="co">## [1] "Metabolism of carbohydrates"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[1]]$miriam.obo.go</span></span>
<span><span class="co">## [1] "GO:0004618"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[1]]$miriam.reactome</span></span>
<span><span class="co">## [1] "REACT_1186" "REACT_5354"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[1]]$miriam.ec</span></span>
<span><span class="co">## [1] "2.7.2.3"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[1]]$miriam.uniprot</span></span>
<span><span class="co">## [1] "P00558"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[1]]$miriam.obo.chebi</span></span>
<span><span class="co">## [1] "CHEBI:18420"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[1]]$miriam.kegg.compound</span></span>
<span><span class="co">## [1] "C00305"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[1]]$compartment.miriam.obo.go</span></span>
<span><span class="co">## [1] "GO:0005829"</span></span></code></pre>
<p>Here, our reaction vertex annotations describing the chemical
transition. You can also notice annotations starting with “miriam” key
word. MIRIAM is a standard format for writing biological identifiers.
You can explore the details of this annotation system on <a href="http://www.ebi.ac.uk/miriam/main/collections" class="external-link uri">http://www.ebi.ac.uk/miriam/main/collections</a>. To list
all vertex attributes:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/getAttr.html">getAttrNames</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "name"                      "reversible"               </span></span>
<span><span class="co">##  [3] "reactants"                 "reactant.stoichiometry"   </span></span>
<span><span class="co">##  [5] "products"                  "product.stoichiometry"    </span></span>
<span><span class="co">##  [7] "kinetics"                  "genes"                    </span></span>
<span><span class="co">##  [9] "compartment"               "compartment.name"         </span></span>
<span><span class="co">## [11] "pathway"                   "miriam.obo.go"            </span></span>
<span><span class="co">## [13] "miriam.reactome"           "miriam.ec"                </span></span>
<span><span class="co">## [15] "miriam.uniprot"            "miriam.obo.chebi"         </span></span>
<span><span class="co">## [17] "miriam.kegg.compound"      "compartment.miriam.obo.go"</span></span></code></pre>
<p>Since annotations tend to be incomplete, NPM provides a function to
check the coverage of each attribute. The function also list the number
of vertices having multiple attribute values. For example, vertices with
multiple <code>miriam.kegg.genes</code> annotations can be view as
protein complexes.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/getAttr.html">getAttrStatus</a></span><span class="op">(</span><span class="va">graph</span>, pattern <span class="op">=</span> <span class="st">"^miriam."</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                      missing single complex coverage.pct</span></span>
<span><span class="co">## miriam.obo.go            269    163       0           38</span></span>
<span><span class="co">## miriam.reactome            0    269     163          100</span></span>
<span><span class="co">## miriam.ec                294    138       0           32</span></span>
<span><span class="co">## miriam.uniprot           184    131     117           57</span></span>
<span><span class="co">## miriam.obo.chebi         230    188      14           47</span></span>
<span><span class="co">## miriam.kegg.compound     289    134       9           33</span></span></code></pre>
<p>NPM also implements an Attribute Fetcher, where you can convert one
annotation to another. The Attribute Fetcher requires RCurl installed,
because it uses the online web service of BridgeDb <a href="http://www.bridgedb.org/" class="external-link uri">http://www.bridgedb.org/</a>.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st">"RCurl"</span><span class="op">)</span></span>
<span><span class="co"># Fetch uniprot annotation</span></span>
<span><span class="va">graph</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MIRIAM.html">fetchAttribute</a></span><span class="op">(</span><span class="va">graph</span>, organism <span class="op">=</span> <span class="st">"Homo sapiens"</span>, target.attr <span class="op">=</span> <span class="st">"miriam.ncbigene"</span> , source.attr <span class="op">=</span> <span class="st">"miriam.uniprot"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Fetch ChEBI annotation.</span></span>
<span><span class="va">graph</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MIRIAM.html">fetchAttribute</a></span><span class="op">(</span><span class="va">graph</span>, target.attr <span class="op">=</span> <span class="st">"miriam.chebi"</span>, source.attr <span class="op">=</span> <span class="st">"miriam.kegg.compound"</span><span class="op">)</span></span></code></pre></div>
<p>You can also use the Attribute Fetcher to obtain Affymetrix
annotation needed for microarray analysis.</p>
</div>
<div class="section level2">
<h2 id="network-processing">Network Processing<a class="anchor" aria-label="anchor" href="#network-processing"></a>
</h2>
<p>NetPathMiner can convert between different network representations.
Given a bipartite metabolic network, reaction network is created by
removing metabolite vertices and keeping them as edge attributes. This
is useful to get adjacent vertices (now reactions) to have gene
annotations needed for gene expression mapping.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rgraph</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/makeReactionNetwork.html">makeReactionNetwork</a></span><span class="op">(</span><span class="va">graph</span>, simplify<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">rgraph</span></span></code></pre></div>
<pre><code><span><span class="co">## IGRAPH b2aceb8 DN-- 180 303 -- </span></span>
<span><span class="co">## + attr: source (g/c), type (g/c), name (v/c), attr (v/x), compound</span></span>
<span><span class="co">## | (e/x), attr (e/x)</span></span>
<span><span class="co">## + edges from b2aceb8 (vertex names):</span></span>
<span><span class="co">##  [1] reaction_71850-&gt;reaction_70420  reaction_71850-&gt;reaction_450088</span></span>
<span><span class="co">##  [3] reaction_71850-&gt;reaction_71541  reaction_71850-&gt;reaction_70467 </span></span>
<span><span class="co">##  [5] reaction_71850-&gt;reaction_70486  reaction_71850-&gt;reaction_163773</span></span>
<span><span class="co">##  [7] reaction_71850-&gt;reaction_70333  reaction_71850-&gt;reaction_71654 </span></span>
<span><span class="co">##  [9] reaction_71850-&gt;reaction_73580  reaction_71850-&gt;reaction_453337</span></span>
<span><span class="co">## [11] reaction_71850-&gt;reaction_70355  reaction_71850-&gt;reaction_174392</span></span>
<span><span class="co">## [13] reaction_71850-&gt;reaction_70349  reaction_71850-&gt;reaction_174389</span></span>
<span><span class="co">## + ... omitted several edges</span></span></code></pre>
<p>Since gene annotations are rarely complete, we can further remove
reaction vertices that are missing gene annotations. This is
particularly meaningful when reactions are translocation or spontaneous
reactions, which are not catalysed by genes. We can then remove such
reactions by:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rgraph</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simplifyReactionNetwork.html">simplifyReactionNetwork</a></span><span class="op">(</span><span class="va">rgraph</span><span class="op">)</span></span>
<span><span class="va">rgraph</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/makeReactionNetwork.html">makeReactionNetwork</a></span><span class="op">(</span><span class="va">graph</span>, simplify<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Some reaction vertices will be catalysed by more than one enzyme. We
can <code>expand</code> these vertices to get the gene network.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Expand complexes of gene network.</span></span>
<span><span class="va">ggraph</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/makeGeneNetwork.html">expandComplexes</a></span><span class="op">(</span><span class="va">rgraph</span>, v.attr <span class="op">=</span> <span class="st">"miriam.uniprot"</span>,</span>
<span>        keep.parent.attr<span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"^pathway"</span>, <span class="st">"^compartment"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Convert reaction network to gene network.</span></span>
<span><span class="va">ggraph</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/makeGeneNetwork.html">makeGeneNetwork</a></span><span class="op">(</span><span class="va">rgraph</span><span class="op">)</span></span></code></pre></div>
<p><code>expandComplexes</code> offers manipulation of network vertices
by their attributes, that includes handling missing annotations and
annotation inheritance. Refer to the manual for for details.</p>
</div>
<div class="section level2">
<h2 id="weighting-network">Weighting Network<a class="anchor" aria-label="anchor" href="#weighting-network"></a>
</h2>
<p>Now that we have our network, we can use gene expression data to
weight the network edges. For this example we use an subset of data
provided by <code>ALL</code> data package. The data consist of
microarrays from 128 different individuals with acute lymphoblastic
leukemia (ALL).</p>
<p>The gene expression is present as Affymetrix IDs. Since we don’t have
these annotations in out network, we can use Attribute Fetcher to get
them.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">ex_microarray</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Assign weights to edges.</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st">"RCurl"</span><span class="op">)</span> <span class="op">&amp;&amp;</span> <span class="fu"><a href="https://rdrr.io/pkg/RCurl/man/url.exists.html" class="external-link">url.exists</a></span><span class="op">(</span> <span class="fu"><a href="../reference/NPMdefaults.html">NPMdefaults</a></span><span class="op">(</span><span class="st">"bridge.web"</span><span class="op">)</span> <span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">rgraph</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MIRIAM.html">fetchAttribute</a></span><span class="op">(</span><span class="va">rgraph</span>, organism <span class="op">=</span> <span class="st">"Homo sapiens"</span>,</span>
<span>                        target.attr <span class="op">=</span> <span class="st">"miriam.affy.probeset"</span>,</span>
<span>                        source.attr <span class="op">=</span> <span class="st">"miriam.uniprot"</span><span class="op">)</span></span></code></pre></div>
<p>Now that we checked that we have <code>affy.probeset</code>
annotations, we can use the weight function. The default weight function
assigns edge weights based on Pearson’s correlation of expression
profiles of adjacent genes. You can also provide you own function as a
<code>weight.method</code>. You can refer to the manual of this function
for details.</p>
<p>We can also provide sample categories as <code>y</code> labels. In
that case, edge weights are computed for each label separately. Here we
use Leukaemia molecular subtypes as categories.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># This requires an internet connection, and RCurl and ALL packages to be present.</span></span>
<span><span class="co"># Instead, we will actually use a processed ALL data, where features are converted</span></span>
<span><span class="co"># to miriam.uniprot annotation. (Next chunk)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">ALL</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">ALL</span><span class="op">)</span></span>
<span><span class="va">rgraph</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/assignEdgeWeights.html">assignEdgeWeights</a></span><span class="op">(</span>microarray <span class="op">=</span> <span class="fu">exprs</span><span class="op">(</span><span class="va">ALL</span><span class="op">)</span>, graph <span class="op">=</span> <span class="va">rgraph</span>,</span>
<span>weight.method <span class="op">=</span> <span class="st">"cor"</span>, use.attr<span class="op">=</span><span class="st">"miriam.affy.probeset"</span>, y<span class="op">=</span><span class="va">ALL</span><span class="op">$</span><span class="va">mol.bio</span>, bootstrap <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 100 genes were present in the microarray, but not represented in the network.</span></span>
<span><span class="co">## 55 genes were couldn't be found in microarray.</span></span>
<span><span class="co">## Assigning edge weights for label ALL1/AF4 </span></span>
<span><span class="co">## Assigning edge weights for label BCR/ABL </span></span>
<span><span class="co">## Assigning edge weights for label E2A/PBX1 </span></span>
<span><span class="co">## Assigning edge weights for label NEG</span></span></code></pre>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rgraph</span><span class="op">$</span><span class="va">y.labels</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "ALL1/AF4" "BCR/ABL"  "E2A/PBX1" "NEG"</span></span></code></pre>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/pkg/igraph/man/E.html" class="external-link">E</a></span><span class="op">(</span><span class="va">rgraph</span><span class="op">)</span><span class="op">$</span><span class="va">edge.weights</span> <span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [[1]]</span></span>
<span><span class="co">## [1] 0.48975367 0.15652908 0.48692055 0.03109587</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[2]]</span></span>
<span><span class="co">## [1] 0.07300260 0.08559941 0.08657545 0.05117537</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[3]]</span></span>
<span><span class="co">## [1] 0.2137908 0.7902335 0.7329898 0.3139489</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[4]]</span></span>
<span><span class="co">## [1] 0.6218986 0.5860552 0.2311897 0.1017890</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[5]]</span></span>
<span><span class="co">## [1] 0.07300260 0.08559941 0.08657545 0.05117537</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[6]]</span></span>
<span><span class="co">## [1]  0.40936252  0.32598614 -0.52609992  0.05701338</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="path-ranking">Path Ranking<a class="anchor" aria-label="anchor" href="#path-ranking"></a>
</h2>
<p>Edges are now weighted by the correlation of connected genes. We can
find highly correlated paths within the network by maximising the edge
weight.</p>
<p>NetPathMiner provides two methods to accomplish that. First,
<code>probabilistic.shortest.path</code> formulates the problem as
finding shortest paths in a network by transforming edge weights by
their empirical cumulative distribution function (ECDF). Finding the
shortest path is equivalent to finding the least probable path given
this empirical distribution. The code below gets the 100-shortest
paths.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ranked.p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pathRanker.html">pathRanker</a></span><span class="op">(</span><span class="va">rgraph</span>, method <span class="op">=</span> <span class="st">"prob.shortest.path"</span>,</span>
<span>    K <span class="op">=</span> <span class="fl">25</span>, minPathSize <span class="op">=</span> <span class="fl">6</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Extracting the 25 most probable paths for ALL1/AF4</span></span></code></pre>
<pre><code><span><span class="co">## Extracting the 25 most probable paths for BCR/ABL</span></span></code></pre>
<pre><code><span><span class="co">## Extracting the 25 most probable paths for E2A/PBX1</span></span></code></pre>
<pre><code><span><span class="co">## Extracting the 25 most probable paths for NEG</span></span></code></pre>
<p>Second, <code>value</code> method finds paths where the sum of edge
weights are significantly higher than random paths of similar length.
The distribution of random path scores can be estimated by
<code>samplePaths</code> which uses Metropolis sampling technique. The
path sample can be then provided to the path ranking function. If path
sample is not provided, random edge sampling is used to estimate the
distribution.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pathsample</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/samplePaths.html">samplePaths</a></span><span class="op">(</span><span class="va">rgraph</span>, max.path.length <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/igraph/man/gorder.html" class="external-link">vcount</a></span><span class="op">(</span><span class="va">rgraph</span><span class="op">)</span>,</span>
<span>num.samples <span class="op">=</span> <span class="fl">1000</span>, num.warmup <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ranked.p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pathRanker.html">pathRanker</a></span><span class="op">(</span><span class="va">rgraph</span>, method <span class="op">=</span> <span class="st">"pvalue"</span>,</span>
<span>sampledpaths <span class="op">=</span> <span class="va">pathsample</span> ,alpha<span class="op">=</span><span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<p>We can get our path set as lists of edge IDs instead.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get paths as edge IDs.</span></span>
<span><span class="va">eids</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getPathsAsEIDs.html">getPathsAsEIDs</a></span><span class="op">(</span>paths <span class="op">=</span> <span class="va">ranked.p</span>, graph <span class="op">=</span> <span class="va">rgraph</span><span class="op">)</span></span></code></pre></div>
<p>We can also get paths as edge IDs on another network representation.
In this example, we extracted paths from a reaction network. We can get
the equivalent paths on the gene network by supplying the corresponding
igraph object.</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Convert paths to other networks.</span></span>
<span><span class="va">eids</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getPathsAsEIDs.html">getPathsAsEIDs</a></span><span class="op">(</span>paths <span class="op">=</span> <span class="va">ranked.p</span>, graph <span class="op">=</span> <span class="va">ggraph</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="clustering-and-classification-of-paths">Clustering and classification of paths<a class="anchor" aria-label="anchor" href="#clustering-and-classification-of-paths"></a>
</h2>
<p>The size of the ranked path can be very large, making their analysis
challenging. NetPathMiner offers clustering functions to group the
ranked path list into few path clusters that can be investigated
easily.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Clustering.</span></span>
<span><span class="va">ybinpaths</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pathsToBinary.html">pathsToBinary</a></span><span class="op">(</span><span class="va">ranked.p</span><span class="op">)</span></span>
<span><span class="va">p.cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pathCluster.html">pathCluster</a></span><span class="op">(</span><span class="va">ybinpaths</span>, M <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotClusters.html">plotClusters</a></span><span class="op">(</span><span class="va">ybinpaths</span>, <span class="va">p.cluster</span><span class="op">)</span></span></code></pre></div>
<p><img src="NPMVignette_files/figure-html/unnamed-chunk-30-1.png" width="700"></p>
<p>We can also identify a set of paths that best classify a sample
category (as a form of biomarker for example). The code below creates a
classifier for <code>BCR/ABL</code> subtype. Since our network is very
small, we are not able to create an accurate classifier.</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p.class</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pathClassifier.html">pathClassifier</a></span><span class="op">(</span><span class="va">ybinpaths</span>, target.class <span class="op">=</span> <span class="st">"BCR/ABL"</span>, M <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotClassifierROC.html">plotClassifierROC</a></span><span class="op">(</span><span class="va">p.class</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="ROCplot.png" title="Logo Title Text 1" alt=""><p class="caption">alt text</p>
</div>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotClusters.html">plotClusters</a></span><span class="op">(</span><span class="va">ybinpaths</span>, <span class="va">p.class</span><span class="op">)</span></span></code></pre></div>
<p><img src="NPMVignette_files/figure-html/unnamed-chunk-33-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="plotting">Plotting<a class="anchor" aria-label="anchor" href="#plotting"></a>
</h2>
<p>NetPathMiner offers several plotting options for networks and ranked
paths. First, we can plot our network colouring vertices by their
cellular compartment.</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotNetwork.html">plotNetwork</a></span><span class="op">(</span><span class="va">rgraph</span>, vertex.color<span class="op">=</span><span class="st">"compartment.name"</span><span class="op">)</span></span></code></pre></div>
<p><img src="NPMVignette_files/figure-html/unnamed-chunk-34-1.png" width="700"></p>
<p>NetPathMiner’s plotPaths function can be used to view ranked paths on
the network structure. Cluster information can also be provided, so that
paths belonging to the same cluster will have the same colour.</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotPaths.html">plotPaths</a></span><span class="op">(</span><span class="va">ranked.p</span>, <span class="va">rgraph</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># With clusters</span></span>
<span><span class="fu"><a href="../reference/plotPaths.html">plotPaths</a></span><span class="op">(</span><span class="va">ranked.p</span>, <span class="va">graph</span>, path.clusters<span class="op">=</span><span class="va">p.class</span><span class="op">)</span></span></code></pre></div>
<p>To view paths on different network representations, you can pass the
networks as parameters to plotPaths, and it will do the job.</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotAllNetworks.html">plotAllNetworks</a></span><span class="op">(</span><span class="va">ranked.p</span>, metabolic.net <span class="op">=</span> <span class="va">graph</span>, reaction.net <span class="op">=</span> <span class="va">rgraph</span>,</span>
<span>        path.clusters<span class="op">=</span><span class="va">p.class</span>, vertex.label <span class="op">=</span> <span class="st">""</span>, vertex.size <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<p><img src="NPMVignette_files/figure-html/unnamed-chunk-36-1.png" width="700"></p>
<p>To make use of the annotation attributes, NetPathMiner can layout
vertices such that those sharing a common attribute value are plotted
close to each other, and using similar colors.</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">layout.c</span> <span class="op">&lt;-</span> <span class="fu">clusterVertexByAttr</span><span class="op">(</span><span class="va">rgraph</span>, <span class="st">"pathway"</span>, cluster.strength <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="va">v.color</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/colorVertexByAttr.html">colorVertexByAttr</a></span><span class="op">(</span><span class="va">rgraph</span>, <span class="st">"pathway"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotPaths.html">plotPaths</a></span><span class="op">(</span><span class="va">ranked.p</span> , <span class="va">rgraph</span>, clusters<span class="op">=</span><span class="va">p.class</span>,</span>
<span>    layout <span class="op">=</span> <span class="va">layout.c</span>, vertex.color <span class="op">=</span> <span class="va">v.color</span><span class="op">)</span></span></code></pre></div>
<p>Finally, for interactive visualization using Cytoscape,
plotCytoscapeGML can export the graph, attributes and layout as a GML
file, which can be imported directly into Cytoscape. For example:</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotCytoscape.html">plotCytoscapeGML</a></span><span class="op">(</span><span class="va">graph</span>, file<span class="op">=</span><span class="st">"example.gml"</span>, layout <span class="op">=</span> <span class="va">layout.c</span>,</span>
<span>                vertex.size <span class="op">=</span> <span class="fl">5</span>, vertex.color <span class="op">=</span> <span class="va">v.color</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="additional-functions">Additional functions<a class="anchor" aria-label="anchor" href="#additional-functions"></a>
</h2>
<div class="section level3">
<h3 id="genesets-and-geneset-subnetworks">Genesets and geneset subnetworks<a class="anchor" aria-label="anchor" href="#genesets-and-geneset-subnetworks"></a>
</h3>
<p>NetPathMiner provides functions to extract genesets utilizing
annotation attributes in the network. To get genesets as lists of genes
for geneset enrichment analyses:</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/getGeneSets.html">getGeneSets</a></span><span class="op">(</span><span class="va">graph</span>, use.attr<span class="op">=</span><span class="st">"compartment"</span>, gene.attr<span class="op">=</span><span class="st">"miriam.uniprot"</span><span class="op">)</span></span></code></pre></div>
<p>Alternatively, genesets can be obtained as network structures.</p>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/getGeneSetNetworks.html">getGeneSetNetworks</a></span><span class="op">(</span><span class="va">graph</span>, use.attr<span class="op">=</span><span class="st">"compartment"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="integration-with-graph-package">Integration with graph package<a class="anchor" aria-label="anchor" href="#integration-with-graph-package"></a>
</h3>
<p>All networks constructed in NetPathMiner are represented as igraph
object. Users can convert these networks to Bioconductor’s graphNEL
object using <code>toGraphNEL</code> function</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">graphNEL</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/toGraphNEL.html">toGraphNEL</a></span><span class="op">(</span><span class="va">graph</span>, export.attr<span class="op">=</span><span class="st">"^miriam."</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Ahmed Mohamed, Tim Hancock, Tim Hancock.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
